File: shared/analysis/lib/qa_analysis/branding.py
"""Product attribution for human-readable reports.

One renderer, one metadata file, three output formats. Every report the pack
produces for a human ends with the same footer, and changing it means editing
`branding.json` — nothing else.

## Why this is code and not a string in each skill

A footer typed by a model is a footer that drifts: the tagline gains a word, the
URL loses a scheme, one report says "Developed by" and the next says "Built by".
Rendering it deterministically makes every report byte-identical, and makes a
change to the branding a one-file edit that CI can verify
(`scripts/check-branding.mjs` fails if any branding string is hardcoded elsewhere).

## What gets a footer, and what must not

Attribution belongs on documents a person reads. It is noise — or worse, a parsing
hazard — anywhere else.

| Branded | Not branded |
| --- | --- |
| HTML reports | JSON and YAML artifacts (`qa-artifacts/*.json`) |
| PDF reports | CLI stdout, progress output, `--json` output |
| Rendered Markdown reports meant for people | Markdown written for machine consumption |
| Generated documentation | Log files, API responses |
| Evaluation and audit report renderings | The system under test, and anything in a user's own source tree |

The rule behind the table: **if a program will parse it, it gets no footer.** A
contract artifact is an interface, and appending prose to an interface breaks it.

## Formats

- `footer_html()` — a `<footer>` element with inline styling (so a standalone
  report file needs no external stylesheet) and the author's site as a link that
  opens in a new tab with `rel="noopener noreferrer"`.
- `footer_markdown()` — a thematic break, the four lines, the site as a link.
- `footer_text()` — rule-separated plain text for PDF writers and terminals-of-last-resort.
  A PDF library that supports hyperlinks should link the URL; one that does not
  renders this as-is, which is why the URL appears in full rather than as anchor text.

Standard library only, like the rest of `qa_analysis`.
"""

import html
import json
import pathlib

_METADATA_PATH = pathlib.Path(__file__).resolve().parent / "branding.json"

# The visual width of the plain-text rules. Wide enough to frame the four lines,
# narrow enough to survive an 80-column terminal or a PDF margin.
_RULE_WIDTH = 60

_ALLOWED_SCHEMES = ("https://", "http://")


class BrandingError(ValueError):
    """Raised when the branding metadata is missing or unusable."""


def metadata():
    """The branding metadata, as a dict. Read fresh so a change needs no reinstall."""
    try:
        with open(_METADATA_PATH, "r", encoding="utf-8") as handle:
            data = json.load(handle)
    except (OSError, json.JSONDecodeError) as exc:
        raise BrandingError(f"could not read branding metadata at {_METADATA_PATH}: {exc}") from exc

    required = ("projectName", "tagline", "author", "website", "attributionPrefix", "authorPrefix")
    missing = [key for key in required if not data.get(key)]
    if missing:
        raise BrandingError(f"branding metadata is missing: {', '.join(missing)}")
    if not data["website"].startswith(_ALLOWED_SCHEMES):
        # A footer is rendered into HTML, so a non-http scheme here would be an
        # injection vector rather than a typo.
        raise BrandingError(
            f"branding website must start with http:// or https://, got {data['website']!r}"
        )
    return data


def _lines(data):
    """The four footer lines, in order, as plain strings."""
    return [
        f"{data['attributionPrefix']} {data['projectName']}",
        data["tagline"],
        f"{data['authorPrefix']} {data['author']}",
        data["website"],
    ]


def footer_text(width=_RULE_WIDTH):
    """Rule-separated plain text. Used for PDF and any non-markup rendering."""
    data = metadata()
    rule = "-" * width
    body = "\n".join(line.center(width).rstrip() for line in _lines(data))
    return f"{rule}\n{body}\n{rule}\n"


def footer_markdown():
    """A thematic break and the four lines, with the site as a link."""
    data = metadata()
    return (
        "---\n\n"
        f"<sub>{data['attributionPrefix']} **{data['projectName']}** — "
        f"{data['tagline']}<br>\n"
        f"{data['authorPrefix']} "
        f"[{data['author']}]({data['website']})</sub>\n"
    )


def footer_html(class_name="qa-pack-attribution"):
    """A self-contained `<footer>`: inline styles, muted, centered, small.

    The author's site opens in a new tab. `rel="noopener noreferrer"` is not
    optional — a report may be opened from anywhere, and a new tab that can reach
    back into `window.opener` is a real hazard.
    """
    data = metadata()
    site = html.escape(data["website"], quote=True)
    project = html.escape(data["projectName"])
    tagline = html.escape(data["tagline"])
    author = html.escape(data["author"])
    attribution_prefix = html.escape(data["attributionPrefix"])
    author_prefix = html.escape(data["authorPrefix"])
    css_class = html.escape(class_name, quote=True)

    return (
        f'<footer class="{css_class}" style="margin-top:2.5rem;padding-top:1rem;'
        'border-top:1px solid rgba(128,128,128,0.25);font-size:0.75rem;line-height:1.6;'
        'color:#6b7280;text-align:center;font-family:system-ui,-apple-system,'
        'Segoe UI,Roboto,sans-serif;">\n'
        f'  <div>{attribution_prefix} <strong>{project}</strong></div>\n'
        f'  <div>{tagline}</div>\n'
        f'  <div>{author_prefix} '
        f'<a href="{site}" target="_blank" rel="noopener noreferrer" '
        'style="color:inherit;text-decoration:underline;">'
        f'{author}</a></div>\n'
        '</footer>\n'
    )


_RENDERERS = {
    "html": footer_html,
    "markdown": footer_markdown,
    "md": footer_markdown,
    "text": footer_text,
    "txt": footer_text,
    "pdf": footer_text,  # what a PDF writer embeds when it cannot render markup
}

FORMATS = ("html", "markdown", "text")


def footer(fmt="text"):
    """Render the footer in `fmt`. Raises BrandingError on an unknown format."""
    renderer = _RENDERERS.get(fmt.lower())
    if renderer is None:
        raise BrandingError(
            f"unknown branding format {fmt!r}; expected one of: {', '.join(FORMATS)}"
        )
    return renderer()


def append_to(document, fmt="markdown"):
    """Return `document` with the footer appended, idempotently.

    Idempotence matters: a report assembled in stages, or regenerated over its own
    output, must not accumulate footers.
    """
    rendered = footer(fmt)
    if rendered.strip() and rendered.strip() in document:
        return document
    separator = "" if document.endswith("\n") else "\n"
    return f"{document}{separator}\n{rendered}"

File: packages/engine/lib/analysis/branding.mjs
// Product attribution for human-readable reports.
//
// One renderer, one metadata file, three output formats. Every report the pack
// produces for a human ends with the same footer, and changing it means editing
// `branding.json` — nothing else.
//
// ## Why this is code and not a string in each skill
//
// A footer typed by a model is a footer that drifts: the tagline gains a word, the
// URL loses a scheme, one report says "Developed by" and the next says "Built by".
// Rendering it deterministically makes every report byte-identical, and makes a
// change to the branding a one-file edit that CI can verify
// (scripts/check-branding.mjs fails if any branding string is hardcoded elsewhere).
//
// ## What gets a footer, and what must not
//
// Attribution belongs on documents a person reads. It is noise — or worse, a
// parsing hazard — anywhere else.
//
//   Branded                          | Not branded
//   ---------------------------------|--------------------------------------------
//   HTML and PDF reports             | JSON and YAML artifacts under qa-artifacts/
//   Markdown meant for people        | Markdown written for a machine to read
//   Generated documentation          | CLI stdout, progress output, --json output
//   Audit and evaluation renderings  | The system under test, the user's own source
//
// The rule behind the table: **if a program will parse it, it gets no footer.** A
// contract artifact is an interface, and appending prose to an interface breaks it.

import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const here = path.dirname(fileURLToPath(import.meta.url));

// One metadata file, beside this module — so it travels with the engine into a
// bundled skill without a second copy to keep in step.
const METADATA_CANDIDATES = [path.join(here, 'branding.json')];

// The visual width of the plain-text rules. Wide enough to frame the four lines,
// narrow enough to survive an 80-column terminal or a PDF margin.
const RULE_WIDTH = 60;

const ALLOWED_SCHEMES = ['https://', 'http://'];

export class BrandingError extends Error {}

/** The branding metadata. Read fresh, so a change needs no reinstall. */
export function metadata() {
  const source = METADATA_CANDIDATES.find((candidate) => fs.existsSync(candidate));
  if (!source) {
    throw new BrandingError(
      `could not find branding metadata (looked in: ${METADATA_CANDIDATES.join(', ')})`,
    );
  }

  let data;
  try {
    data = JSON.parse(fs.readFileSync(source, 'utf8'));
  } catch (error) {
    throw new BrandingError(`could not read branding metadata at ${source}: ${error.message}`);
  }

  const required = ['projectName', 'tagline', 'author', 'website', 'attributionPrefix', 'authorPrefix'];
  const missing = required.filter((key) => !data[key]);
  if (missing.length > 0) {
    throw new BrandingError(`branding metadata is missing: ${missing.join(', ')}`);
  }
  if (!ALLOWED_SCHEMES.some((scheme) => data.website.startsWith(scheme))) {
    // A footer is rendered into HTML, so a non-http scheme here would be an
    // injection vector rather than a typo.
    throw new BrandingError(
      `branding website must start with http:// or https://, got '${data.website}'`,
    );
  }
  return data;
}

/** HTML-escape, matching Python's `html.escape(s, quote=True)` character for character. */
function escape(value) {
  return String(value)
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#x27;');
}

/** The four footer lines, in order, as plain strings. */
function lines(data) {
  return [
    `${data.attributionPrefix} ${data.projectName}`,
    data.tagline,
    `${data.authorPrefix} ${data.author}`,
    data.website,
  ];
}

/**
 * Rule-separated plain text, for PDF and any non-markup rendering.
 *
 * Centering follows Python's `str.center`: the odd space goes on the right, which
 * then falls off with the trailing trim. Getting that backwards shifts every line
 * by a character and breaks the snapshot the branding tests pin.
 */
export function footerText({ width = RULE_WIDTH } = {}) {
  const data = metadata();
  const rule = '-'.repeat(width);
  const body = lines(data)
    .map((line) => {
      if (line.length >= width) return line;
      return ' '.repeat(Math.floor((width - line.length) / 2)) + line;
    })
    .join('\n');
  return `${rule}\n${body}\n${rule}\n`;
}

/** A thematic break and the four lines, with the site as a link. */
export function footerMarkdown() {
  const data = metadata();
  return (
    '---\n\n' +
    `<sub>${data.attributionPrefix} **${data.projectName}** — ` +
    `${data.tagline}<br>\n` +
    `${data.authorPrefix} ` +
    `[${data.author}](${data.website})</sub>\n`
  );
}

/**
 * A self-contained `<footer>`: inline styles, muted, centered, small.
 *
 * The author's site opens in a new tab. `rel="noopener noreferrer"` is not
 * optional — a report may be opened from anywhere, and a new tab that can reach
 * back into `window.opener` is a real hazard.
 */
export function footerHtml({ className = 'qa-pack-attribution' } = {}) {
  const data = metadata();
  return (
    `<footer class="${escape(className)}" style="margin-top:2.5rem;padding-top:1rem;` +
    'border-top:1px solid rgba(128,128,128,0.25);font-size:0.75rem;line-height:1.6;' +
    'color:#6b7280;text-align:center;font-family:system-ui,-apple-system,' +
    'Segoe UI,Roboto,sans-serif;">\n' +
    `  <div>${escape(data.attributionPrefix)} <strong>${escape(data.projectName)}</strong></div>\n` +
    `  <div>${escape(data.tagline)}</div>\n` +
    `  <div>${escape(data.authorPrefix)} ` +
    `<a href="${escape(data.website)}" target="_blank" rel="noopener noreferrer" ` +
    'style="color:inherit;text-decoration:underline;">' +
    `${escape(data.author)}</a></div>\n` +
    '</footer>\n'
  );
}

const RENDERERS = {
  html: footerHtml,
  markdown: footerMarkdown,
  md: footerMarkdown,
  text: footerText,
  txt: footerText,
  pdf: footerText, // what a PDF writer embeds when it cannot render markup
};

export const FORMATS = ['html', 'markdown', 'text'];

/** Render the footer in `format`. Throws BrandingError on an unknown format. */
export function footer(format = 'text') {
  const renderer = RENDERERS[String(format).toLowerCase()];
  if (!renderer) {
    throw new BrandingError(
      `unknown branding format '${format}'; expected one of: ${FORMATS.join(', ')}`,
    );
  }
  return renderer();
}

/**
 * Return `document` with the footer appended, idempotently.
 *
 * Idempotence matters: a report assembled in stages, or regenerated over its own
 * output, must not accumulate footers.
 */
export function appendTo(document, format = 'markdown') {
  const rendered = footer(format);
  if (rendered.trim() && document.includes(rendered.trim())) return document;
  const separator = document.endsWith('\n') ? '' : '\n';
  return `${document}${separator}\n${rendered}`;
}
